62.CAD/Polyline.h

```diff
@@ -255,10 +281,223 @@ class CPolyline
+					phaseShiftTotal = linePoint.phaseShift;
+				}
+			}
+			else if (section.type == ObjectType::QUAD_BEZIER)
+			{
+				// calculate phase shift at point P0 of each bezier
+				const uint32_t quadBezierCnt = section.count;
+				const bool isLastSection = (sectionIdx == m_sections.size() - 1u);
+
+				for (uint32_t i = 0u; i <= quadBezierCnt; i++)
+				{
+					const bool isLastBezierInSection = (i == quadBezierCnt);
+					// there is no need to calculate anything for the last bezier in the last section
+					if (isLastSection && isLastBezierInSection)
+					{
+						break;
+					}
+
+					const uint32_t currIdx = section.index + i;
+					if (i == 0u)
+					{
+						QuadraticBezierInfo& firstInSectionQuadBezierInfo = m_quadBeziers[currIdx];
+						firstInSectionQuadBezierInfo.phaseShift = phaseShiftTotal;
+
+						if (lineStyle.isRoadStyleFlag)
+						{
+							connectorBuilder.addBezierNormals(m_quadBeziers[currIdx], phaseShiftTotal);
+						}
+
+						continue;
+					}
+
+					const QuadraticBezierInfo& prevQuadBezierInfo = m_quadBeziers[currIdx - 1u];
+					shapes::QuadraticBezier<double> quadraticBezier = shapes::QuadraticBezier<double>::construct(prevQuadBezierInfo.p[0], prevQuadBezierInfo.p[1], prevQuadBezierInfo.p[2]);
+					shapes::Quadratic<double> quadratic = shapes::Quadratic<double>::constructFromBezier(quadraticBezier);
+					shapes::Quadratic<double>::ArcLengthCalculator arcLenCalc = shapes::Quadratic<double>::ArcLengthCalculator::construct(quadratic);
+					const double bezierLen = arcLenCalc.calcArcLen(1.0f);
+
+					const double nextLineInSectionLocalPhaseShift = std::remainder(bezierLen, 1.0f / lineStyle.reciprocalStipplePatternLen) * lineStyle.reciprocalStipplePatternLen;
+					phaseShiftTotal = glm::fract(phaseShiftTotal + nextLineInSectionLocalPhaseShift);
+
+					if (i < quadBezierCnt)
+					{
+						QuadraticBezierInfo& quadBezierInfo = m_quadBeziers[currIdx];
+						quadBezierInfo.phaseShift = phaseShiftTotal;
+
+						if (lineStyle.isRoadStyleFlag)
+						{
+							connectorBuilder.addBezierNormals(m_quadBeziers[currIdx], phaseShiftTotal);
+						}
+					}
+				}
+			}
+		}
+
+		if (lineStyle.isRoadStyleFlag)
+		{
+			m_polylineConnector = connectorBuilder.buildConnectors();
+		}
+	}
+
 protected:
 	// TODO[Przemek]: a vector of polyline connetor objects
+	std::vector<PolylineConnector> m_polylineConnector;
 	std::vector<SectionInfo> m_sections;
-	// TODO[Przemek]: instead of float64_t2 for linePoints, store LinePointInfo
-	std::vector<float64_t2> m_linePoints;
+	std::vector<LinePointInfo> m_linePoints;
 	std::vector<QuadraticBezierInfo> m_quadBeziers;
+
+private:
+	class PolylineConnectorBuilder
+	{
+	public:
+		void addLineNormal(const float64_t2& line, float lineLen, const float64_t2& worldSpaceCircleCenter, float phaseShift)
+		{
+			PolylineConnectorNormalHelperInfo connectorNormalInfo;
+			connectorNormalInfo.worldSpaceCircleCenter = float32_t2(worldSpaceCircleCenter);
```

Comment:
reminder: you're throwing out precision here, later on storing this in a float64_t won't make you go back in time.